File: bumperbot_navigation/src/ball_follower.cpp
#include "bumperbot_navigation/ball_follower.hpp"

#include "cv_bridge/cv_bridge.hpp"

#include "opencv2/core.hpp"
#include "opencv2/opencv.hpp"

namespace bumperbot_navigation
{
BallFollower::BallFollower(const std::string & name) : Node(name)
{
  declare_parameter("lower_hsv", std::vector<int64_t>{79, 50, 50});
  declare_parameter("upper_hsv", std::vector<int64_t>{99, 255, 255});
  declare_parameter("ball_radius", 0.31);
  get_parameter("lower_hsv", lower_hsv_);
  get_parameter("upper_hsv", upper_hsv_);
  get_parameter("ball_radius", ball_radius_);

  approximate_sync_ = std::make_shared<ApproximateSync>(
    ApproximatePolicy(10), image_sub_, info_sub_);
  approximate_sync_->registerCallback(
    std::bind(&BallFollower::imageCallback, this, _1, _2));
  image_transport::TransportHints hints{this};
  const auto sensor_data_qos = rclcpp::SensorDataQoS().get_rmw_qos_profile();

  image_sub_.subscribe(
    this, "left_camera/image_rect", hints.getTransport(), sensor_data_qos);
  info_sub_.subscribe(this, "left_camera/camera_info", sensor_data_qos);

  detection_pub_ =
    image_transport::create_publisher(this, "left_camera/ball_detection", rmw_qos_profile_sensor_data);
  ball_pose_pub_ = create_publisher<geometry_msgs::msg::PoseStamped>("goal_update", 10);
}

void BallFollower::imageCallback(
  const sensor_msgs::msg::Image::ConstSharedPtr & image_msg,
  const sensor_msgs::msg::CameraInfo::ConstSharedPtr & info_msg)
{
  cv_bridge::CvImagePtr cv_ptr;
  try {
    cv_ptr = cv_bridge::toCvCopy(image_msg, sensor_msgs::image_encodings::BGR8);
  } catch (cv_bridge::Exception & e) {
    RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
    return;
  }
  
  cv::Mat image = cv_ptr->image;
  cv::Mat hsv, mask;

  cv::cvtColor(image, hsv, cv::COLOR_BGR2HSV);

  cv::Scalar lower_bound(lower_hsv_[0], lower_hsv_[1], lower_hsv_[2]);
  cv::Scalar upper_bound(upper_hsv_[0], upper_hsv_[1], upper_hsv_[2]);
  cv::inRange(hsv, lower_bound, upper_bound, mask);

  cv::erode(mask, mask, cv::Mat(), cv::Point(-1, -1), 2);
  cv::dilate(mask, mask, cv::Mat(), cv::Point(-1, -1), 2);

  std::vector<std::vector<cv::Point>> contours;
  cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

  if (!contours.empty()) {
    auto largest_contour = std::max_element(contours.begin(), contours.end(),
      [](const std::vector<cv::Point>& a, const std::vector<cv::Point>& b) {
        return cv::contourArea(a) < cv::contourArea(b);
      });

    cv::Point2f center;
    float radius;
    cv::minEnclosingCircle(*largest_contour, center, radius);

    if (radius > 10.0) {
      cv::Rect bbox = cv::boundingRect(*largest_contour);
      cv::rectangle(image, bbox, cv::Scalar(0, 255, 0), 2);
      cv::circle(image, center, 3, cv::Scalar(0, 0, 255), -1);

      double fx = info_msg->k[0];
      double cx = info_msg->k[2];
      double fy = info_msg->k[4];
      double cy = info_msg->k[5];

      if (fx > 0 && fy > 0) {
        geometry_msgs::msg::PoseStamped pose_msg;
        pose_msg.header.stamp = image_msg->header.stamp;
        pose_msg.header.frame_id = info_msg->header.frame_id;

        double estimated_z = (fx * ball_radius_) / radius;
        double estimated_x = (center.x - cx) * estimated_z / fx;
        double estimated_y = (center.y - cy) * estimated_z / fy;

        pose_msg.pose.position.x = estimated_x;
        pose_msg.pose.position.y = estimated_y;
        pose_msg.pose.position.z = estimated_z;
        pose_msg.pose.orientation.w = 1.0; 

        ball_pose_pub_->publish(pose_msg);
      }
    }
  }

  sensor_msgs::msg::Image::SharedPtr detection_msg = cv_bridge::CvImage(
    image_msg->header, sensor_msgs::image_encodings::BGR8, image).toImageMsg();
  detection_pub_.publish(*detection_msg);
}
}  // namespace bumperbot_navigation

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<bumperbot_navigation::BallFollower>("ball_follower"));
  rclcpp::shutdown();
  return 0;
}

File: bumperbot_navigation/bumperbot_navigation/ball_follower.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
import message_filters
from sensor_msgs.msg import Image, CameraInfo
from geometry_msgs.msg import PoseStamped
from cv_bridge import CvBridge

import cv2
import numpy as np

class BallFollower(Node):
    def __init__(self):
        super().__init__('ball_follower')
        
        # 1. Declare Node Parameters with Gazebo defaults
        self.declare_parameter('lower_hsv', [79, 50, 50])
        self.declare_parameter('upper_hsv', [99, 255, 255])
        self.declare_parameter('ball_radius', 0.31)
        self.lower_hsv = self.get_parameter('lower_hsv').value
        self.upper_hsv = self.get_parameter('upper_hsv').value
        self.ball_radius = self.get_parameter('ball_radius').value
        
        self.bridge = CvBridge()
        
        self.image_sub = message_filters.Subscriber(
            self, Image, 'left_camera/image_rect', qos_profile=qos_profile_sensor_data)
        self.info_sub = message_filters.Subscriber(
            self, CameraInfo, 'left_camera/camera_info', qos_profile=qos_profile_sensor_data)
            
        self.ts = message_filters.ApproximateTimeSynchronizer(
            [self.image_sub, self.info_sub], queue_size=10, slop=0.1)
        self.ts.registerCallback(self.image_callback)
        
        self.detection_pub = self.create_publisher(
            Image, 'left_camera/ball_detection', qos_profile_sensor_data)
        self.ball_pose_pub = self.create_publisher(
            PoseStamped, 'goal_update', 10)

    def image_callback(self, image_msg, info_msg):
        
        # Convert ROS Image to OpenCV format
        try:
            cv_image = self.bridge.imgmsg_to_cv2(image_msg, desired_encoding='bgr8')
        except Exception as e:
            self.get_logger().error(f"CvBridge Error: {e}")
            return
            
        # Convert to HSV
        hsv_image = cv2.cvtColor(cv_image, cv2.COLOR_BGR2HSV)
        
        # Create NumPy arrays for the color bounds
        lower_bound = np.array(self.lower_hsv)
        upper_bound = np.array(self.upper_hsv)
        
        # Create the mask
        mask = cv2.inRange(hsv_image, lower_bound, upper_bound)
        
        # Morphological Operations to clean up noise and shadows
        kernel = np.ones((5, 5), np.uint8)
        # OPENING: Remove small background noise
        mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel, iterations=2)
        # CLOSING: Fill in holes and gaps caused by shadows
        mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel, iterations=5)
        
        # Find contours
        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        
        if contours:
            # Find the largest contour by area
            largest_contour = max(contours, key=cv2.contourArea)
            
            # Get the minimum enclosing circle
            ((x, y), radius) = cv2.minEnclosingCircle(largest_contour)
            
            if radius > 10.0:
                # Draw the bounding box and center
                x_int, y_int = int(x), int(y)
                x_rect, y_rect, w, h = cv2.boundingRect(largest_contour)
                
                cv2.rectangle(cv_image, (x_rect, y_rect), (x_rect + w, y_rect + h), (0, 255, 0), 2)
                cv2.circle(cv_image, (x_int, y_int), 3, (0, 0, 255), -1)
                
                # Pose Estimation using CameraInfo
                fx = info_msg.k[0]
                cx = info_msg.k[2]
                fy = info_msg.k[4]
                cy = info_msg.k[5]
                
                if fx > 0 and fy > 0:
                    pose_msg = PoseStamped()
                    pose_msg.header.stamp = image_msg.header.stamp
                    pose_msg.header.frame_id = info_msg.header.frame_id
                    
                    # Estimate Z (Depth) and X, Y
                    estimated_z = (fx * self.ball_radius) / radius
                    estimated_x = (x - cx) * estimated_z / fx
                    estimated_y = (y - cy) * estimated_z / fy
                    
                    pose_msg.pose.position.x = float(estimated_x)
                    pose_msg.pose.position.y = float(estimated_y)
                    pose_msg.pose.position.z = float(estimated_z)
                    pose_msg.pose.orientation.w = 1.0
                    
                    self.ball_pose_pub.publish(pose_msg)
                    
        # Convert annotated image back to ROS and publish
        detection_msg = self.bridge.cv2_to_imgmsg(cv_image, encoding="bgr8")
        detection_msg.header = image_msg.header
        self.detection_pub.publish(detection_msg)

def main(args=None):
    rclpy.init(args=args)
    node = BallFollower()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()